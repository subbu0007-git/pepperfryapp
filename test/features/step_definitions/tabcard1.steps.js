import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import { defineFeature, loadFeature } from "jest-cucumber";
import { Getcard1Data } from "../../../Redux/Api/Getcard1data";
import Tabcard1 from "../../../components/Tabcard1";
// Mocking React Navigation's useRoute hook
jest.mock('@react-navigation/native', () => ({
    ...jest.requireActual('@react-navigation/native'),
    useRoute: () => ({
        params: {
            screen: 'mocked-screen', // Provide a default screen name
            // Add any other params you need for testing
        },
    }),
}));

const feature = loadFeature('test/features/tabcard1.feature')

jest.mock('../../../Redux/Api/Getcard1data');
defineFeature(feature, test => {
    let store;
    let screen;
    let setCard1dataMock;
    const mockData = [
        { id: 1, imgurl: 'https://example.com/image1.jpg', status: 'status1' },
        { id: 2, imgurl: 'https://example.com/image2.jpg', status: 'status2' }
    ];
    // 1 first test case : checking the get api call after that based on the routing the data will be rendered in the flatlist
    test('API call and data filtering', ({ given, when, then }) => {
        beforeEach(() => {
            const mockStore = configureStore([]);
            store = mockStore({ Card1reducer: { card1data: mockData } });
        });
        given('I am on the Tabcard1 component', () => {
            Getcard1Data.mockResolvedValue(mockData);
            setCard1dataMock = jest.fn();
            screen = render(
                <Provider store={store}>
                    <Tabcard1
                        setCard1data={setCard1dataMock}
                    />
                </Provider>
            );
        });
        when('the get API call takes place', async () => {
            // No need to mock GetsliderData again here
            await waitFor(() => expect(Getcard1Data).toHaveBeenCalled());
        });
        then('the data should be filtered based on route name', () => {
            const expectedRoute = 'status1'; // Define the expected route name here
            const filteredData = mockData.filter(item => item.status === expectedRoute);
            expect(filteredData.length).toBeGreaterThan(0); // Assert that filtered data is not empty
        });
        then('the FlatList should render the filtered data', () => {
            const renderedData = screen.getAllByTestId('flatlist-item'); // Assuming each item in FlatList has testID 'flatlist-item'
            expect(renderedData.length).toBeGreaterThan(0); // Assert that rendered data is not empty
        });
    });


    // second test case : testing weather the api is catching the error whrn the api is not working
    test('Catching and logging error if get API call fails', ({ given, when, then }) => {
        let consoleErrorSpy;
        beforeEach(() => {
            const mockStore = configureStore([]);
            store = mockStore({ Card1reducer: { card1data: mockData } });
        });
        given('the getTabcard1 API call fails', () => {
            Getcard1Data.mockRejectedValueOnce(new Error('API call failed'));
        });
        when('the Form component is rendered', async () => {
            consoleErrorSpy = jest.spyOn(console, 'error');
            render(
                <Provider store={store}>
                    <Tabcard1 />
                </Provider>
            );
            await waitFor(() => new Promise(resolve => setTimeout(resolve, 100)));
        });
        then('the error should be caught and logged', () => {
            expect(Getcard1Data).toHaveBeenCalled();
            expect(consoleErrorSpy).toHaveBeenCalledWith('An error occurred:', new Error('API call failed'));
        });
    });

    // third test case: testing that renderitem is renderingthe items accurately or not based on the index numbers
    test('Checking the flatlist data is rendering correctly or not', ({ when, then }) => {
        let mockDataForFlatList = [
            { id: 1, imgurl: 'https://example.com/image1.jpg', status: 'status1' },
            { id: 2, imgurl: 'https://example.com/image2.jpg', status: 'status2' }
        ];
        when('the Tabcard1 component is rendered', () => {
            const mockStore = configureStore([]);
            store = mockStore({ Card1reducer: { card1data: mockDataForFlatList } });
            screen = render(
                <Provider store={store}>
                    <Tabcard1 />
                </Provider>
            );
        });
        then('the FlatList should be rendered with correct data', () => {
            const flatList = screen.getByTestId('flatlist-item');
            expect(flatList).toBeTruthy();

            // Ensure that the keyExtractor function is covered
            const keyExtractorProp = flatList.props.keyExtractor;
            expect(keyExtractorProp).toBeTruthy();

            // Simulate key extraction for each item
            const keys = mockDataForFlatList.map(item => keyExtractorProp(item));
            console.log('i am keys',keys)
            // Ensure that the keys are correct
            expect(keys).toEqual([1, 2]);

            // Ensure that the renderItem function is covered
            const renderItemProp = flatList.props.renderItem;
            expect(renderItemProp).toBeTruthy();

            // Simulate rendering items
            const renderedItem1 = renderItemProp({ item: mockDataForFlatList[0], index: 0 });
            const renderedItem2 = renderItemProp({ item: mockDataForFlatList[1], index: 1 });

            // Ensure that the rendered items contain necessary components like Text with correct content
            expect(renderedItem1).toMatchSnapshot();
            expect(renderedItem2).toMatchSnapshot();
        });

    });
});