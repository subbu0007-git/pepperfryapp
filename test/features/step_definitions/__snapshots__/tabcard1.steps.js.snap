// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabcard1 Component Checking the flatlist data is rendering correctly or not 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      {
        "marginRight": 6,
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://example.com/image1.jpg",
        }
      }
      style={
        {
          "height": 216,
          "resizeMode": "contain",
          "width": 140,
        }
      }
    />
  </View>
</View>
`;

exports[`Tabcard1 Component Checking the flatlist data is rendering correctly or not 2`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <React.Fragment>
    <View>
      <Image
        source={
          {
            "uri": "https://example.com/image2.jpg",
          }
        }
        style={
          {
            "height": 160,
            "marginRight": 1,
            "resizeMode": "contain",
            "width": 150,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-end",
            "bottom": 0,
            "justifyContent": "flex-end",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "borderRadius": 25,
              "padding": 5,
              "right": 12,
              "top": 13,
            }
          }
        >
          <Icon
            allowFontScaling={false}
            color="#696969"
            name="heart-outline"
            size={21}
          />
        </View>
      </View>
    </View>
    <View>
      <Text
        ellipsizeMode="tail"
        numberOfLines={2}
        style={
          {
            "color": "#363636",
            "fontSize": 12,
            "fontWeight": "500",
            "marginTop": 7,
            "maxWidth": 139,
          }
        }
      >
        Milan Fabric 3 Seater Sofa in Sky Blue Colour
      </Text>
    </View>
    <View>
      <Text
        style={
          {
            "marginTop": 1,
          }
        }
      >
        <Text
          style={
            {
              "fontWeight": "bold",
            }
          }
        >
          ₹12,499 
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 12,
              "fontWeight": "500",
              "textDecorationLine": "line-through",
            }
          }
        >
           ₹17,000
        </Text>
        <Text
          style={
            {
              "color": "green",
              "fontSize": 14.5,
              "fontWeight": "bold",
            }
          }
        >
           72%
        </Text>
      </Text>
    </View>
  </React.Fragment>
</View>
`;
